sherpa/cpp_api/brpc/server.cc

```diff
@@ -0,0 +1,380 @@
+    }
+  }
+
+  int64_t acquire_stream() {
+    int64_t stream_id = -1;
+    std::shared_lock mlock(mutex_);
+    if (stream_queue_.size() <= max_batch_size_) {
+      while (true) {
+        stream_id = rand() % max_batch_size_;  //NOLINT
+        if (stream_queue_.find(stream_id) == stream_queue_.end()) {
+          break;
+        }
+      }
+      stream_queue_[stream_id] = online_asr_.CreateStream();
+      result_queue_[stream_id] = std::vector<char>(MAX_ASR_RESULT);
+      vad_state_queue_[stream_id] = input_start;
+    } else {
+      SHERPA_LOG(WARNING) << "stream_queue_.size() and max_batch_size_ = "
+        << max_batch_size_ << ", can't create more stream";
+    }
+    return stream_id;
+  }
+
+  void release_stream(const int64_t stream_id) {
+    std::shared_lock mlock(mutex_);
+    if (stream_queue_.find(stream_id) != stream_queue_.end()) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+      stream_queue_.erase(stream_id);
+      result_queue_.erase(stream_id);
+      vad_state_queue_.erase(stream_id);
+    }
+  }
+
+  void reset_stream(const int64_t stream_id) {
+    std::shared_lock mlock(mutex_);
+    if (stream_queue_.find(stream_id) != stream_queue_.end()) {
+      stream_queue_[stream_id] = online_asr_.CreateStream();
+      result_queue_[stream_id] = std::vector<char>(MAX_ASR_RESULT);
+      vad_state_queue_[stream_id] = input_start;
+    }
+  }
+
+  void stop_stream(const int64_t stream_id) {
+    std::shared_lock mlock(mutex_);
+    if (stream_queue_.find(stream_id) != stream_queue_.end()) {
+      vad_state_queue_[stream_id] = input_end;
+    }
+  }
+
+  void push_stream(const int64_t stream_id,
+      const char * data, const size_t len, bool end = false) {
+    std::shared_lock mlock(mutex_);
+    if (stream_queue_.find(stream_id) != stream_queue_.end()) {
+      {
+        if (vad_state_queue_[stream_id] == endpoint_reset) {
+          stream_queue_[stream_id] = online_asr_.CreateStream();
+          vad_state_queue_[stream_id] = endpoint_inactive;
+        }
+        const int num_samples = len / sizeof(int16_t);
+        char * pcm_data = const_cast<char *>(data);
+        auto wav_stream_tensor = torch::from_blob(
+            reinterpret_cast<int16_t *>(pcm_data),
+            {num_samples},
+            torch::kInt16).to(torch::kFloat) / 32768;
+
+        stream_queue_[stream_id]->AcceptWaveform(SAMPLE_RATE,
+            wav_stream_tensor);
+        if (end || vad_state_queue_[stream_id] == input_end) {
+          stream_queue_[stream_id]->InputFinished();
+          vad_state_queue_[stream_id] = endpoint_reset;
+        }
+      }
+      mlock.unlock();
+      batch_cond_.notify_one();
+      if (vad_state_queue_[stream_id] != endpoint_reset) {
+        if (segment_stream(stream_id)) {
+          vad_state_queue_[stream_id] = endpoint_active;
+        }
+      }
+    }
+  }
+
+  const char * fetch_stream(const int64_t stream_id) {
```

Comment:
Since it requires c++17, I suggest that you use a return value of type `std::optional<std::string>`.